bitshuffle/bitshuffle.py

```diff
@@ -316,33 +317,58 @@ def decode(message):
         packets = [re.sub("|".join(string.whitespace), "", p)
                    for p in packets if p.strip() is not '']
 
+        num_chunks_wrong = 0
+
         segments = [None] * len(packets)  # ordered by index of packets
         # each chunk will be appended and original will be returned
         payload = bytes()
+        overall_hash = None
         for index, packet in enumerate(packets):
             try:
-                segments[index] = packet.split("|")
-                if segments[index][seq_num] != str(index):
+                packet = packet.split("|")
+                if packet[seq_num] != str(index):
                     stderr.write("WARNING: Sequence number " +
                                  "%s does not match actual order %d\n"
-                                 % (segments[index][seq_num], index))
+                                 % (packet[seq_num], index))
                     continue
             except IndexError:
                 stderr.write("WARNING: Packet " +
                              "%d is invalid for decoding.\n" %
                              (index))
                 continue
 
-            payload += base64.b64decode(segments[index][chunk])
+            if len(packet) - 1 == file_hash:
+                if not overall_hash:
+                    overall_hash = packet[file_hash]
+                elif packet[file_hash] != overall_hash:
+                    stderr.write(
+                        "WARNING: File hash mismatch in packet" + index)
+
+            hashed = hash(packet[chunk].encode(encoding='ascii'))
+            if hashed != packet[packet_hash]:
+                num_chunks_wrong += 1
+                stderr.write("WARNING: Given hash for packet "
+                             + "%d does not match actual hash '%s'\n"
+                             % (index, packet[packet_hash]))
+
+            payload += base64.b64decode(packet[chunk])
 
-        if segments[0][compression] == "bz2":
+        if packet[compression] == "bz2":
             payload = bz2.decompress(payload)
         else:
             payload = gzip_decompress(payload)
 
-        checksum_ok = verify(payload, segments[0][checksum])
-        payload = bytes(payload)
-        return payload, checksum_ok
+        file_hash_ok = (num_chunks_wrong == 0
```

Comment:
This smells like it might not quite do what you intend. Are you sue `overall_hash` is a boolean?

I'm not sure I follow `or (overall_hash and hash(payload) == overall_hash))`.

Reply:
Since python has [type-coercion](https://en.wikipedia.org/wiki/Type_conversion),
`if <non-bool>:` or any sort of boolean comparison is the same as `if bool(<non-bool>)`. 

` overall_hash and hash(payload) == overall_hash` is true IF the overall_hash has been defined AND is equal to the hash of the concatenated data.

```python
>>> bool
<class 'bool'>
>>> bool(None)
False
>>> if None:
...     print('true')
... 
>>> 
```

Reply:
Ahh, that makes sense, but is difficult to read at first glance. I would consider instead writing something like `(overall_hash is not None) and (hash(payload) == overall_hash)`. While longer, I think it's more obvious what you are actually trying to accomplish.

Reply:
Done in https://github.com/charlesdaniels/bitshuffle/pull/46/commits/b07624365f66b3ae5c5a0124186a1b216c66a500

Reply:
Looks good, thank you!